osquery/remote/http/http_client.cpp

```diff
@@ -0,0 +1,342 @@
+    ctx.set_verify_mode(boost_asio::ssl::verify_peer);
+  } else {
+    ctx.set_verify_mode(boost_asio::ssl::verify_none);
+  }
+
+  if (client_options_.server_certificate_) {
+    ctx.set_verify_mode(boost_asio::ssl::verify_peer);
+    ctx.load_verify_file(*client_options_.server_certificate_);
+  }
+
+  if (client_options_.verify_path_) {
+    ctx.set_verify_mode(boost_asio::ssl::verify_peer);
+    ctx.add_verify_path(*client_options_.verify_path_);
+  }
+
+  if (client_options_.ciphers_) {
+    ::SSL_CTX_set_cipher_list(ctx.native_handle(),
+                              client_options_.ciphers_->c_str());
+  }
+
+  if (client_options_.ssl_options_) {
+    ctx.set_options(client_options_.ssl_options_);
+  }
+
+  if (client_options_.client_certificate_file_) {
+    ctx.use_certificate_file(*client_options_.client_certificate_file_,
+                             boost_asio::ssl::context::pem);
+  }
+
+  if (client_options_.client_private_key_file_) {
+    ctx.use_private_key_file(*client_options_.client_private_key_file_,
+                             boost_asio::ssl::context::pem);
+  }
+
+  ssl_sock_ = std::make_shared<ssl_stream>(sock_, ctx);
+  if (client_options_.sni_hostname_) {
+    ::SSL_set_tlsext_host_name(ssl_sock_->native_handle(),
+                               client_options_.sni_hostname_->c_str());
+  }
+
+  boost_system::error_code rc;
+  ssl_sock_->handshake(boost_asio::ssl::stream_base::client, rc);
+
+  if (rc) {
+    throw std::system_error(rc.value(), adapted_category(&rc.category()));
+  }
+}
+
+template <typename STREAM_TYPE>
+void Client::sendRequest(STREAM_TYPE& stream,
+                         Request& req,
+                         beast_http_response_parser& resp) {
+  req.target((req.remotePath()) ? *req.remotePath() : "/");
+  req.version = 11;
+  req << Request::Header("Host", *client_options_.remote_hostname_);
+  req.prepare_payload();
+  req.keep_alive(true);
+
+  if (client_options_.timeout_) {
+    timer_.async_wait(
+        [=](boost_system::error_code const& ec) { timeoutHandler(ec); });
+  }
+
+  beast_http_request_serializer sr{req};
+  beast_http::async_write(stream, sr, [&](boost_system::error_code const& ec) {
+    if (client_options_.timeout_) {
+      timer_.cancel();
+    }
+
+    if ((ec.value() != boost_system::errc::operation_canceled) ||
+        (ec.category() != boost_asio::error::system_category)) {
+      ec_ = ec;
+    }
+  });
+
+  ios_.run();
+  ios_.reset();
+
+  if (ec_) {
+    throw std::system_error(ec_.value(), adapted_category(&ec_.category()));
+  }
+
+  if (client_options_.timeout_) {
+    timer_.async_wait(
+        [=](boost_system::error_code const& ec) { timeoutHandler(ec); });
+  }
+
+  boost::beast::flat_buffer b;
+  beast_http::async_read(
+      stream, b, resp, [&](boost_system::error_code const& ec) {
+        if (client_options_.timeout_) {
+          timer_.cancel();
+        }
+        postResponseHandler(ec);
+      });
+
+  ios_.run();
```

Comment:
FYI: If an asynchronous operation throws an exception, the `io_service` will not be reset.

Reply:
Two reasons - 
1. In our usage we use the client object once. 
2. I find there is no reason for this to throw the exception in this context.

I will keep this in mind once we enhance the http_client class in the future.